man/man4/memcached.json.4
'\" t
.\"     Title: memcached.json
.\"    Author: Trond Norbye <trond.norbye@couchbase.com>
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 03/27/2015
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "MEMCACHED\&.JSON" "4" "03/27/2015" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
memcached.json \- memcached configuration file
.SH "DESCRIPTION"
.sp
\fBmemcached\&.json\fR is a JSON encoded file specifying the properties used to configure the memcached server\&. Some of the parameters may be changed at runtime by instructing memcached to reread the configuration file\&. These properties is explicitly marked as such\&.
.sp
The following sections describes the various attributes that may be specified\&.
.SS "root"
.sp
This is the root directory of the Couchbase server installation\&.
.SS "admin"
.sp
Specify the username used for administrive operations (like bucket management)\&. To disable the use of an administrative role you should set this to an empty string\&.
.SS "breakpad"
.sp
The \fBbreakpad\fR attribute is used to configure the Breakpad crash catcher\&. When enabled (and on a supported platform), if memcached crashes a minidump containing information on the state of memcached will be written to disk\&. It is an object with the following attributes:
.sp
.if n \{\
.RS 4
.\}
.nf
enabled       A boolean value specifying if Breakpad is enabled\&.
              If true (and *minidump_dir* is set) minidumps will
              be output to directory specified by *minidump_dir*\&.
              If not specified then defaults to false\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
minidump_dir  A string value specifying the directory to write any
              outputted minidumps to\&.  If not specified then
              Breakpad is not enabled\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
content       A string value specifying what data will be included
              in generated minidumps\&. Currently the only permitted
              value is "default"\&.
.fi
.if n \{\
.RE
.\}
.sp
\fBenabled\fR, \fBminidump_dir\fR and \fBcontent\fR may be modified at runtime by instructing memcached to reread the configuration file\&.
.SS "require_init"
.sp
The \fBrequire_init\fR attribute is a boolean value that is used to disable disable all user commands while the server (Couchbase Server) is initializing the node (creating the buckets etc)\&. Until the node is initialized memcached will only allow the "admin user" to connect to the cluster and run commands\&. All other users will receive a "NOT INITIALIZED" response for all commands except SASL requests; which will be allowed, but upon a successful authentication "NOT INITIALIZED" will be returned unless the SASL authentication was done for the admin user\&.
.SS "rbac_file"
.sp
Specify the filename containing all of the RBAC definitions
.SS "rbac_privilege_debug"
.sp
The \fBrbac_privilege_debug\fR is a boolean value that may be used by developers to detect the privileges required for their module to work\&. It should \fInever\fR be enabled in a production environment!
.sp
When running in the privilege debug mode, all commands will be permitted for all users\&. If the current user profile lacks the permission for the operation to be performed, the following entry is printed to standard error:
.sp
Missing privilege for \fBuser profile\fR\&. Need \fBcommand\fR\&.
.SS "audit_file"
.sp
Specify the filename containing all of the Audit configurations
.SS "ssl_cipher_list"
.sp
Specify the list of ciphers to use\&. The format of the string is described in ciphers ( https://www\&.openssl\&.org/docs/apps/ciphers\&.html )\&. By default no value is specified causing whatever openssl defaults to to be enabled\&. Some values may be:
.sp
.if n \{\
.RS 4
.\}
.nf
HIGH          Only enable "high" encrypthion ciphers (keys > 128
              bits)
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
MEDIUM        In addition to the high encryption ciphers, also
              allow 128 bit encryption
.fi
.if n \{\
.RE
.\}
.SS "threads"
.sp
The \fBthreads\fR attribute specify the number of threads used to serve clients\&. By default this number is set to 75% of the number of cores available on the system (but no less than 4)\&. The value for threads should be specified as an integral number\&.
.SS "interfaces"
.sp
The \fBinterfaces\fR attribute is used to specify an array of interfaces memcached should listen at\&. Each entry in the interfaces array is an object describing a single interface with the following properties:
.sp
.if n \{\
.RS 4
.\}
.nf
host          A string value specifying the hostname to bind to\&.
              If the attribute is missing (or set to "*")
              IN_ADDR_ANY is used\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
port          An integral number specifying the port number
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
IPv4          A boolean value specifying if IPv4 should be enabled
              or not\&. By default IPv4 is enabled\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
IPv6          A boolean value specifying if IPv6 should be enabled
              or not\&. By default IPv6 is enabled\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
maxconn       An integral number specifying the maximum number of
              concurrent clients that may be connected\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
backlog       An integral number specifying the backlog size
              set to listen\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
tcp_nodelay   A boolean value if TCP_NODELAY should be set or not\&.
              By default tcp_nodelay is enabled\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
ssl           An object specifying SSL related properties\&.
              See below\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
protocol      A string value specifying the protocol enabled
              for this port\&. If not present the memcached binary
              protocol is used\&. Legal values: "greenstack" or
              "memcached"
.fi
.if n \{\
.RE
.\}
.sp
The \fBssl\fR object contains the two \fBmandatory\fR attributes:
.sp
.if n \{\
.RS 4
.\}
.nf
key           A string value with the absolute path to the
              file containing the private key to use\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
cert          A string value with the absolute path to the
              file containing the X\&.509 certificate to use\&.
.fi
.if n \{\
.RE
.\}
.sp
\fBmaxconn\fR, \fBbacklog\fR, \fBtcp_nodelay\fR, \fBssl\&.key\fR and \fBssl\&.cert\fR may be modified by instructing memcached to reread the configuration file\&.
.SS "extensions"
.sp
The \fBextensions\fR attribute is used to specify an array of extensions which should be loaded\&. Each entry in the extensions array is an object describing a single extension with the following attributes:
.sp
.if n \{\
.RS 4
.\}
.nf
module    A string value containing the module to load\&. This may
          either be an absolute path, or the systems library path
          will be searched for the object\&. When loaded memcached
          calls the method memcached_extensions_initialize()\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
config    A string value containing configuration attributes to
          the module\&. The config value is passed transparently
          to the module\&.
.fi
.if n \{\
.RE
.\}
.SS "engine"
.sp
The \fBengine\fR attribute is used to specify the engine to load\&. It is an object with the following attributes:
.sp
.if n \{\
.RS 4
.\}
.nf
module    A string value containing the module to load\&. This may
          either be an absolute path, or the systems library path
          will be searched for the object\&. The object must
          implement the engine api\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
config    A string value containing configuration attributes to
          the module\&. The config value is passed transparently
          to the module\&.
.fi
.if n \{\
.RE
.\}
.SS "require_sasl"
.sp
The \fBrequire_sasl\fR attribute specify if performing SASL authentication is required or not\&. The value is a boolean value which is set to false by default\&.
.SS "default_reqs_per_event"
.sp
The \fBdefault_reqs_per_event\fR attribute is an integral value specifying the number of request that may be served per client before serving the next client (to avoid starvation)\&. The default value is 20\&.
.sp
\fBdefault_reqs_per_event\fR may be updated by instructing memcached to reread the configuration file\&.
.SS "reqs_per_event_high_priority"
.sp
The \fBreqs_per_event_high_priority\fR attribute is an integral value specifying the number of request that may be served per high priority client before serving the next client (to avoid starvation)\&. The default value is 20\&.
.sp
\fBreqs_per_event_high_priority\fR may be updated by instructing memcached to reread the configuration file\&.
.SS "reqs_per_event_med_priority"
.sp
The \fBreqs_per_event_med_priority\fR attribute is an integral value specifying the number of request that may be served per medium priority client before serving the next client (to avoid starvation)\&. The default value is 20\&.
.sp
\fBreqs_per_event_med_priority\fR may be updated by instructing memcached to reread the configuration file\&.
.SS "reqs_per_event_low_priority"
.sp
The \fBreqs_per_event_low_priority\fR attribute is an integral value specifying the number of request that may be served per low priority client before serving the next client (to avoid starvation)\&. The default value is 20\&.
.sp
\fBreqs_per_event_low_priority\fR may be updated by instructing memcached to reread the configuration file\&.
.SS "bio_drain_buffer_sz"
.sp
The \fBbio_drain_buffer_sz\fR attribute is an integral value specifying the number of bytes in the BIO drain buffer\&. This is an interal setting just used by the engineers for testing\&.
.SS "verbosity"
.sp
The \fBverbosity\fR attribute is an integral value specifying the amount of output produced by the memcached server\&. By default this value is set to 0 resulting in only warnings to be emitted\&. Setting this value too high will produce a lot of output which is most likely meaningless for most people\&.
.sp
\fBverbosity\fR may be updated by instructing memcached to reread the configuration file\&.
.SS "datatype_support"
.sp
The \fBdatatype_support\fR attribute is a boolean value to enable the support for using the datatype extension\&. By default this support is \fBdisabled\fR\&.
.SS "max_packet_size"
.sp
The \fBmax_packet_size\fR attribute is an integer value that specify the maximum packet size (in MB) allowed to be received from clients without disconnecting them\&. This is a safetynet for avoiding the server to try to spool up a 4GB packet\&. When a packet is received on the network with a body bigger than this threshold EINVAL is returned to the client and the client is disconnected\&.
.SH "EXAMPLES"
.sp
A Sample memcached\&.json:
.sp
.if n \{\
.RS 4
.\}
.nf
{
    "root" : "/opt/couchbase",
    "admin" : "_admin",
    "breakpad" :
        {
            "enabled" : true,
            "minidump_dir" : "/opt/couchbase/var/crash",
            "content" : "default"
        },
    "require_init" : true,
    "rbac_file" : "/opt/couchbase/etc/security/rbac\&.json",
    "rbac_privilege_debug" : false,
    "audit_file" : "/opt/couchbase/etc/security/audit\&.json",
    "ssl_cipher_list" : "HIGH",
    "threads" : 4,
    "interfaces" :
    [
        {
            "host" : "*",
            "port" : 11209,
            "IPv4" : true,
            "IPv6" : true,
            "maxconn" : 10000,
            "backlog" : 1024,
            "tcp_nodelay" : true,
            "ssl" :
            {
                "key" : "/etc/memcached/pkey",
                "cert" : "/etc/memcached/cert"
            },
            "protocol" : "memcached"
        },
        {
            "host" : "*",
            "port" : 11210,
            "IPv4" : true,
            "IPv6" : true,
            "maxconn" : 10000,
            "backlog" : 1024,
            "tcp_nodelay" : true,
            "ssl" :
            {
                "key" : "/etc/memcached/pkey",
                "cert" : "/etc/memcached/cert"
            },
            "protocol" : "greenstack"
        }
    ],
    "extensions" :
    [
        {
            "module" : "stdin_term_handler\&.so",
            "config" : ""
        },
        {
            "module" : "file_logger\&.so",
            "config" : "cyclesize=10485760;sleeptime=19;filename=data/n_0/logs/memcached\&.log"
        }
    ],
    "engine" : {
        "module" : "bucket_engine\&.so",
        "config" : "admin=_admin;default_bucket_name=default;auto_create=false"
    },
    "require_sasl" : false,
    "default_reqs_per_event" : 20,
    "reqs_per_event_high_priority" : 40,
    "reqs_per_event_med_priority" : 20,
    "reqs_per_event_low_priority" : 10,
    "verbosity" : 2,
    "datatype_support" : true
    "max_packet_size" : 25,
    "bio_drain_buffer_sz" : 8192
}
.fi
.if n \{\
.RE
.\}
.SH "COPYRIGHT"
.sp
Copyright 2015 Couchbase, Inc\&.
.SH "AUTHOR"
.PP
\fBTrond Norbye\fR <\&trond\&.norbye@couchbase\&.com\&>
.RS 4
Author.
.RE


memcached.json.ex
{
    "threads" : 4,
    "interfaces" :
    [
        {
            "maxconn" : 10000,
            "port" : 11210,
            "backlog" : 1024,
            "host" : "*",
            "IPv6" : true,
            "ssl" :
            {
                "key" : "/etc/memcached/pkey",
                "cert" : "/etc/memcached/cert"
            }
        },
        {
            "maxconn" : 1000,
            "port" : 11213,
            "host" : "127.0.0.1",
            "IPv6" : false
        }
    ],
    "extensions" :
    [
        {
            "module" : "stdin_term_handler.so",
            "config" : ""
        },
        {
            "module" : "file_logger.so",
            "config" : "cyclesize=10485760;sleeptime=19;filename=data/n_0/logs/memcached.log"
        }
    ],
    "engine" : {
        "module" : "bucket_engine.so",
        "config" : "admin=_admin;default_bucket_name=default;auto_create=false"
    },
    "require_sasl" : false,
    "prefix_delimiter" : ":",
    "tcp_nodelay" : false,
    "allow_detailed" : true,
    "detail_enabled" : false,
    "reqs_per_event" : 20,
    "verbosity" : 0,
    "lock_memory" : false,
    "large_memory_pages" : false,
    "daemonize" : false,
    "pid_file" : "/var/run/memcached.pid"
}
